include/device.h

```diff
@@ -732,10 +732,21 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
 		(&DEVICE_NAME_GET(dev_name)), level, prio)
 
 #ifdef CONFIG_PM_DEVICE
-#define Z_DEVICE_DEFINE_PM_INIT(dev_name, pm_control_fn)		\
+#define Z_DEVICE_STATE_DEFINE(node_id, dev_name)			\
+	static struct device_state Z_DEVICE_STATE_NAME(dev_name) = {	\
+		.pm = {						\
+			.flags = ATOMIC_INIT(DT_PROP_OR(		\
+				node_id, wakeup_source, 0) <<		\
```

Comment:
Just use DT_PROP.

Reply:
nope, this does not work when something that expand to an invalid node. e.g: ``SYS_DEVICE_DEFINE("sys_clock"```

Reply:
> nope, this does not work when something that expand to an invalid node. e.g: ``SYS_DEVICE_DEFINE("sys_clock"```

Then I think the caller can do

```
IF_ENABLED(DT_NODE_EXISTS(node_id), (Z_DEVICE_STATE_DEFINE(node_id, dev_name))) \
```

or so

Reply:
That implies in changes that are beyond this patch IMHO. The state has to be defined for cases where a device is not defined using DT. Of course, I can just do `IF_ENABLED(DT_NOT_EXISTS(node_id), ...) ` in that place, but it makes it much more verbose. Do you think is really to bad just keep this way ?

Reply:
I agree with @mbolivar-nordic, and actually the same technique seems to be used for the device handles, see https://github.com/zephyrproject-rtos/zephyr/blob/0d3c75c2b09f9d11c04f52a92c5fb0fbef2ea32b/include/device.h#L696

Another change request I missed: we have `Z_DEVICE_DEFINE_PRE` for all stuff that needs to be defined before device, so I think the state definition also belongs to that group.

Reply:
>  Do you think is really to bad just keep this way ?

Yes, I am pushing everyone to use DT_PROP() instead of DT_NODE_HAS_PROP() when it comes to boolean properties for a reason. The reason is that using DT_PROP() means you benefit from the property system and type checking in the bindings language. If you use DT_NODE_HAS_PROP() instead, and you misspell the property, you can get an incorrect result and it would be very hard to debug. If you make a typo on the property name using DT_PROP(), then you get a build error instead. So using DT_NODE_HAS_PROP() for booleans is an antipattern I'm willing to block on.

Reply:
the device handles is not doing the same thing, it is not accessing properties. Using `COND_CODE_1(DT_NODE_EXISTS(node_id), () ()` there is just more cumbersome than using `DT_PROP_OR` IMHO, since they are doing exactly the same thing. Now ... if you are saying to *NOT* define the state struct in case the device is not defined using DT that is not right because this struct is required for devices that are defined manually using `DEVICE_DEFINE`.

So in the end, it seems just a matter of changing  `DT_PROP_OR` with `COND_CODE_1(DT_NODE_EXISTS(node_id) ...`. I particularly prefer the first, it looks cleaner in my opinion, but I am ok changing it.

Regards `Z_DEVICE_DEFINE_PRE`, that makes sense.